### PR TITLE
Variables: Fixes Textbox current value persistence

### LINKED ISSUE #28991: Variables: Textbox variable always updates the default value when saving the dashboard
**What happened**:

When saving a dashboard, the value of variables are saved as dashboard default values, regardless the checkbox `Save current variable values as dashboard default` is chosen or not.

**What you expected to happen**:

Will only save as default when chosen.

**How to reproduce it (as minimally and precisely as possible)**:

* create a dashboard with variables
* change the value in the dashboard
* save without activating the checkbox
* reload or logout/login

**Anything else we need to know?**:

**Environment**: 
- Grafana version: 7.3.1
- Data source type & version: influxdb 1.8
- OS Grafana is installed on: Docker official image

### PR DESCRIPTION
**Pull request by hugohaggmark**
**What this PR does / why we need it**:
This PR is a branch off from #29407 and tries to make persistence for text box variables clearer.

**Which issue(s) this PR fixes**:
Fixes #28991

**Special notes for your reviewer**:

### CODE REVIEW

File: public/app/features/dashboard/state/DashboardMigrator.test.ts [hunk 1]
```diff
@@ -798,6 +799,107 @@ describe('DashboardModel', () => {
...
+              options: [{ selected: true, text: 'A', value: 'B' }],
+              datasource: null,
+              allFormat: '',
+            },
+          ],
+        },
+      });
+    });
+
+    it('should have three variables after migration', () => {
```
Comment (dprokop) [inline public/app/features/dashboard/state/DashboardMigrator.test.ts:860]:
```suggestion
    it('should have six variables after migration', () => {
```

File: public/app/features/variables/textbox/TextBoxVariablePicker.tsx [hunk 2]
```diff
@@ -1,43 +1,47 @@
...
+
+  const onBlur = useCallback((e: FocusEvent<HTMLInputElement>) => updateVariable(), [updateVariable, updatedValue]);
+  const onKeyDown = useCallback(
+    (event: KeyboardEvent<HTMLInputElement>) => {
+      if (event.keyCode === 13) {
+        updateVariable();
+      }
+    },
+    [updateVariable, updatedValue]
+  );
```
Comment (dprokop) [inline public/app/features/variables/textbox/TextBoxVariablePicker.tsx:44]: Not sure we need to memoize this handler, as the useCallback will be invalidated with any `updatedValue` change anyways. WDYT @hugohaggmark ?

File: public/app/features/variables/textbox/TextBoxVariablePicker.tsx [hunk 3]
```diff
@@ -1,43 +1,47 @@
...
-      dispatch(updateOptions(toVariableIdentifier(this.props.variable)));
-    }
-  };
+  const dispatch = useDispatch();
+  const [updatedValue, setUpdatedValue] = useState(value);
 
-  onQueryKeyDown = (event: KeyboardEvent<HTMLInputElement>) => {
-    if (event.keyCode === 13 && this.props.variable.current.value !== this.props.variable.query) {
-      dispatch(updateOptions(toVariableIdentifier(this.props.variable)));
+  const updateVariable = useCallback(() => {
```
Comment (dprokop) [inline public/app/features/variables/textbox/TextBoxVariablePicker.tsx:23]: Same comment about memoization as the one about `onKeyDown` handler :)

**Review by dprokop — APPROVED** [review]
Great stuff @hugohaggmark! Left minor nits to consider :)

File: public/app/features/variables/textbox/TextBoxVariablePicker.tsx (hunk 2 above)
Comment (hugohaggmark) [reply public/app/features/variables/textbox/TextBoxVariablePicker.tsx:44]: Yes, good catch!

File: public/app/features/variables/textbox/TextBoxVariablePicker.tsx (hunk 3 above)
Comment (hugohaggmark) [reply public/app/features/variables/textbox/TextBoxVariablePicker.tsx:20]: Fixed it, thank you!

**Pull request merged**